File: src/leetcode/m173/BSTIterator.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.m173;

import leetcode.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * LeetCode 173
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/10/10 09:44
 */
public class BSTIterator {
    private final List<Integer> list;
    private int next = 0;

    public BSTIterator(TreeNode root) {
        this.list = convertToList(root);
    }

    public int next() {
        return list.get(next++);
    }

    public boolean hasNext() {
        return next < list.size();
    }

    private List<Integer> convertToList(TreeNode root) {
        if (root.left == null && root.right == null) {
            List<Integer> res = new ArrayList<>();
            res.add(root.val);
            return res;
        }
        if (root.left == null) {
            List<Integer> rightList = convertToList(root.right);
            List<Integer> res = new ArrayList<>();
            res.add(root.val);
            res.addAll(rightList);
            return res;
        }
        if (root.right == null) {
            List<Integer> leftList = convertToList(root.left);
            List<Integer> res = new ArrayList<>(leftList);
            res.add(root.val);
            return res;
        }
        List<Integer> leftList = convertToList(root.left);
        List<Integer> rightList = convertToList(root.right);
        List<Integer> res = new ArrayList<>(leftList);
        res.add(root.val);
        res.addAll(rightList);
        return res;
    }
}

File: src/leetcode/m173/BSTIterator.go
package m173

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// BSTIterator LeetCode 173
type BSTIterator struct {
	slice []int
}

func Constructor(root *TreeNode) BSTIterator {
	iterator := BSTIterator{
		slice: make([]int, 0),
	}
	build(root, &iterator.slice)
	return iterator
}

func build(root *TreeNode, slice *[]int) {
	if root == nil {
		return
	}
	build(root.Left, slice)
	*slice = append(*slice, root.Val)
	build(root.Right, slice)
}

func (this *BSTIterator) Next() int {
	res := this.slice[0]
	this.slice = this.slice[1:]
	return res
}

func (this *BSTIterator) HasNext() bool {
	return len(this.slice) > 0
}
